### PR TITLE
database

### PR DESCRIPTION
**Pull request by leo9070**
basic of database libs

### CODE REVIEW

File: kujira/database/mongodb/lib/mongodb_lib.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        self.client = MongoClient()
+        self.database_connection = None
+        self.database_collection = None
+        self.database_collection_history = None
+        self.post = None
+        self.document = None
+
+    def connect(self, data_base_name, collection_name, history_collection_name):
+        '''method setting up the connection to database'''
+        self.database_connection = self.client[data_base_name]
```
Comment (niekrasp) [inline kujira/database/mongodb/lib/mongodb_lib.py]: attribute name 'connection' is enough

File: kujira/database/mongodb/lib/mongodb_lib.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        '''method inserting task(actually just a name) as record into specified
+         collections in database, auto generated date, to allow getting oldest event'''
+        self.post = {"name": name, "date": datetime.datetime.utcnow()}
+        return self.database_collection.insert_one(self.post).inserted_id
+
+
+    def get_task(self):
+        '''gets the newest task from database, deletes it and then inserts into
+        historical collection with beginning state "doing"'''
+        temporary_cursor = (self.database_collection.find().sort("date", 1).limit(1))
```
Comment (niekrasp) [inline kujira/database/mongodb/lib/mongodb_lib.py]: 1. get_task() should just return first (last inserted) task.
2. new method to update status of task 'set_status'
3. new method close_task to add task to tasks_history collection, set its status and result, then remove task from tasks collection

File: kujira/database/mongodb/lib/mongodb_lib.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        self.database_connection = None
+        self.database_collection = None
+        self.database_collection_history = None
+        self.post = None
+        self.document = None
+
+    def connect(self, data_base_name, collection_name, history_collection_name):
+        '''method setting up the connection to database'''
+        self.database_connection = self.client[data_base_name]
+        self.database_collection = self.database_connection[collection_name]
```
Comment (niekrasp) [inline kujira/database/mongodb/lib/mongodb_lib.py]: self.tasks_collection

File: kujira/database/mongodb/lib/mongodb_lib.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        self.database_collection_history = self.database_connection[history_collection_name]
```
Comment (niekrasp) [inline kujira/database/mongodb/lib/mongodb_lib.py]: self.tasks_audit_collection

File: kujira/database/mongodb/lib/mongodb_lib.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+
+    def insert_task(self, name):
```
Comment (niekrasp) [inline kujira/database/mongodb/lib/mongodb_lib.py]: task has name, but also arguments, this should be dictionary

File: kujira/database/mongodb/lib/mongodb_lib.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        temporary_cursor = (self.database_collection.find().sort("date", 1).limit(1))
+        self.document = self.database_collection.find_one({"_id": temporary_cursor[0]['_id']})
+        self.post = {"_id":self.document['_id'],
+                     "name":self.document['name'],
+                     "date": datetime.datetime.utcnow(), "state": "doing"}
+        self.database_collection_history.insert_one(self.post)
+        self.database_collection.delete_one({"_id": self.document['_id']})
+        return self.document
+
+    def update_evaluated_task(self, task_id, new_state):
```
Comment (niekrasp) [inline kujira/database/mongodb/lib/mongodb_lib.py]: remove it, close_task does the job

File: kujira/database/mongodb/lib/mongodb_lib.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
+'''
```
Comment (niekrasp) [inline kujira/database/mongodb/lib/mongodb_lib.py]: Change file name and path to: kujira/store/tasks.py

File: kujira/database/redis/lib/redis_lib.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
+'''module docString'''
```
Comment (niekrasp) [inline kujira/database/redis/lib/redis_lib.py]: Change path and name to 'kujira/store/events_queue.py'

File: kujira/database/redis/lib/redis_lib.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+import redis
+
+class Redis(object):
+    '''class '''
+    def __init__(self):
+        '''init'''
```
Comment (niekrasp) [inline kujira/database/redis/lib/redis_lib.py]: Configuration file !!!

File: kujira/database/redis/lib/redis_lib.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        self.connection = False
+        self.connection_adress = 'localhost'
+        self.connection_port = 6379
+        self.redis_connection = None
+
+    def connect(self):
+        '''connecting'''
+        self.redis_connection = redis.Redis(self.connection_adress)
```
Comment (niekrasp) [inline kujira/database/redis/lib/redis_lib.py]: Here should be line empty

File: kujira/database/mongodb/lib/mongodb_lib.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+library to acces Mongo database allowing to insert and get tasks.
```
Comment (niekrasp) [inline kujira/database/mongodb/lib/mongodb_lib.py]: Where is error handling?

File: kujira/database/redis/lib/redis_lib.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    def push_event_to_queue(self, event):
+        '''pushes event to redis database into list'''
+        self.redis_connection.lpush('event_queue', event)
```
Comment (niekrasp) [inline kujira/database/redis/lib/redis_lib.py]: connection is enough

File: kujira/database/redis/lib/redis_lib.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+    def pop_event_from_queue(self):
```
Comment (niekrasp) [inline kujira/database/redis/lib/redis_lib.py]: pop is enough as method name

File: kujira/database/redis/lib/redis_lib.py [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+        '''pops oldest event in redis list, inf there is none, waits till any appear'''
+        return self.redis_connection.brpop('event_queue').replace("\'", "\"")
+
+    def is_any_event_in_queue(self):
```
Comment (niekrasp) [inline kujira/database/redis/lib/redis_lib.py]: is_not_empty()

File: kujira/database/redis/lib/redis_lib.py [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+        '''checks if there is any event in list'''
+        return bool(self.redis_connection.llen('event_queue') > 0)
+
+    def is_connected(self):
+        '''checks connections'''
+        return bool(self.redis_connection.ping())
+
+    def push_token(self, user_name, random_value):
```
Comment (niekrasp) [inline kujira/database/redis/lib/redis_lib.py]: Extract token management to separate module 'kujira/storage/tokens.py'

File: kujira/store/events_queue.py [hunk 16]
```diff
@@ -0,0 +1,42 @@
+'''events queue module'''
+from . import redis_db
+
+
+class RedisQueue(redis_db. RedisConnection):
+    '''management of tasks in Redis database '''
+    def __init__(self):
+        '''init'''
+        super(RedisQueue, self).__init__()
+        self.connection_adress = 'localhost'
```
Comment (tlna) [inline kujira/store/events_queue.py]: Cosntructor of base class should set attributes: connection_address and connection_port

File: kujira/store/redis_db.py [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+        '''connecting'''
+        self.redis_connection = redis.Redis(self.connection_adress)
+
+    @abstractmethod
+    def is_connected(self):
+        '''checks connections'''
+        try:
+            return bool(self.redis_connection.ping())
+        except redis.ConnectionError:
+            raise Exception('Cannot connect to database!')
```
Comment (tlna) [inline kujira/store/redis_db.py]: Create module kujira/store/exceptions.py and define you custom exceptions there.
Do not use Exception to throw/catch exceptions

File: kujira/store/tasks.py [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+import datetime
+from pymongo import MongoClient
+from pymongo import errors
+
+class Mongodb(object):
+    '''contains connecting, inserting and geting tasks from store'''
+
+    def __init__(self):
+        self.client = MongoClient()
+        self.connection_r = None
```
Comment (tlna) [inline kujira/store/tasks.py]: What is connection_r?

File: kujira/store/tasks.py [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+        self.tasks_collection = None
+        self.tasks_audit_collection = None
+        self.post = None
```
Comment (tlna) [inline kujira/store/tasks.py]: what is post?

File: kujira/store/tasks.py [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+        self.document = None
```
Comment (tlna) [inline kujira/store/tasks.py]: Why do you need document here?

File: kujira/store/tasks.py [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+
+    def connect(self, data_base_name, collection_name, history_collection_name):
+        '''method setting up the connection to store'''
+        self.connection_r = self.client[data_base_name]
+        self.tasks_collection = self.connection_r[collection_name]
+        self.tasks_audit_collection = self.connection_r[history_collection_name]
+
+    def insert_task(self, param_dict):
+        '''method inserting task(actually just a name) as record into specified
```
Comment (tlna) [inline kujira/store/tasks.py]: Use double quotes '" for docstrings.

File: kujira/store/tasks.py [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+        except IndexError:
+            raise Exception('Cursor empty, problem with database!')
+
+    def get_task(self):
+        '''gets the newest task from store, deletes it and then inserts into
+        historical collection with beginning state "doing"'''
+        try:
+            temporary_cursor = (self.tasks_collection.find().sort("date", 1).limit(1))
+            self.document = self.tasks_collection.find_one({"_id": temporary_cursor[0]['_id']})
+            self.tasks_collection.update_one(
```
Comment (tlna) [inline kujira/store/tasks.py]: get_task should only return it, not modify

File: kujira/store/tasks.py [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+                {"_id": self.document["_id"]},
+                {"$set": {"task_state": "doing"}}
+                )
+            return self.document
+        except errors.ConnectionFailure:
+            raise Exception('Cannot connect to database!')
+        except IndexError:
+            raise Exception('Cursor empty, problem with database!')
+
+    def close_task(self, task_id):
```
Comment (tlna) [inline kujira/store/tasks.py:55]: Add error handling. Method should be atomic.

**Pull request merged**